### ISSUE TITLE
Conjugate gradient Descent, and Linear operator are not present in pytorch.

### ISSUE THREAD

**Issue by NeelMishra** [pytorch/pytorch#53441]
## 🚀 Feature

Conjugate gradient descent, and Linear operator as implemented in scipy needs to have a place in pytorch for faster gpu calculations.

## Motivation

Conjugate gradient Descent and Linear operator are a powerful tool for solving linear equations which many of the optimizer uses. Currently we have to work in scipy, and shift the vectors between numpy and pytorch to utilize the autograd function of pytorch. 

We can try cupy for faster access but again cupy isn't supported in pytorch, so it increases additional dependencies namely, converting between cupy, numpy, and pytorch tensors.

## Pitch

Adding linear Operator, and then conjugate gradient method next is a great way to improve pytorch. These are the only reasons I am using Scipy, it would be a great way for welcoming scipy community in pytorch too.

**Comment by anjali411** [comment]
Hi @NeelMishra! Thanks for filing this issue. Are you referring to these sparse operations: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.cg.html https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html?

**Comment by NeelMishra** [comment]
@anjali411 Yes, those are exactly it. It will be great help to me as well as other researchers to try different algorithm with the help of conjugate gradient descent.

**Comment by anjali411** [comment]
@NeelMishra I see that scipy implements conjugate gradient method iteratively instead of using the direct method (ref: https://en.wikipedia.org/wiki/Conjugate_gradient_method). what are the pros and cons of both approaches and do you have a preference of one over the other? We do have a solve function in PyTorch https://pytorch.org/docs/stable/generated/torch.solve.html?highlight=solve#torch.solve can that be useful while we don't have support for CG in PyTorch?

**Comment by NeelMishra** [comment]
Sorry, I am new to research so I don't know the pros and cons of conjugate gradient method. But I am very much sure that iterative CG is way better than closed form/direct solution in the fields of deep learning, where closed form computations can be way costly, and almost not always required. 

It is indeed very helpful to quick test, and see if a proposed algorithm is working, ex a proposed optimizer's update rule can be written as a system of linear equation there in CG plays a critical role, and Linear operator is just a very good complementary addition when the entries of the A matrix are not entirely known.

**Comment by anjali411** [comment]
> a proposed optimizer's update rule can be written as a system of linear equation there in CG plays a critical role, and Linear operator is just a very good complementary addition when the entries of the A matrix are not entirely known.

Makes sense!

**Comment by IvanYashchuk** [comment]
> Linear operator is just a very good complementary addition when the entries of the A matrix are not entirely known.

When the entries of the A matrix are not known you can't define a linear operator. To clarify a [LinearOperator from SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html) (or anything equivalent) is useful in cases when it's known how to define a matrix-vector multiplication without explicit instantiation of the matrix. It's in the sparse namespace because it's often used together with sparse matrices, but it is not constrained only to sparse matrices. In general, methods that do not require an explicit matrix but only its matrix-vector product are called [matrix-free](https://en.wikipedia.org/wiki/Matrix-free_methods). SciPy's LinearOperator can be directly used with PyTorch and no need to replicate this class; to construct a concrete LinearOperator, either pass appropriate callables to the constructor of the class or subclass it.

SciPy's conjugate gradient internally calls NumPy, so PyTorch tensors, even under LinearOperator, can't work. However, the conjugate gradient is simple to implement in PyTorch as it's just a sequence of matrix-vector multiplications, and the [algorithm is concise](https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method).

I think this functionality can well exist in a separate external package. For example SciPy's documentation for LinearOperator suggests [PyLops](https://github.com/PyLops/pylops) project:
> More details regarding how to subclass a LinearOperator and several examples of concrete LinearOperator instances can be found in the external project [PyLops](https://github.com/PyLops/pylops).

There is [PyLops-GPU](https://pylops-gpu.readthedocs.io/en/latest/index.html) that uses PyTorch and has the [conjugate gradient solver implemented](https://pylops-gpu.readthedocs.io/en/latest/_modules/pylops_gpu/optimization/cg.html#cg).

**Comment by anjali411** [comment]
> There is [PyLops-GPU](https://pylops-gpu.readthedocs.io/en/latest/index.html) that uses PyTorch and has the [conjugate gradient solver implemented](https://pylops-gpu.readthedocs.io/en/latest/_modules/pylops_gpu/optimization/cg.html#cg).

I think this is ok but not a great option since the `ComplexTensor` wrapper used under the hood uses the `(..., 2)` floating point tensor to represent complex data and the operators defined on these [`ComplexTensors`](https://openbase.com/python/pytorch-complex-tensor/documentation), unlike PyTorch, don't utilize specialized complex kernels for operations like multiplication where using the  native complex tensors gives significantly better performance. I'll file an issue on `PyLops` in case they'd like to switch to native complex tensors.

**Comment by Balandat** [comment]
There is #28341 about this that has been open for a while. 

We have a basic implementation in https://github.com/cornellius-gp/linear_operator, but so far it's focused on symmetric positive semidefinite operators (at least for now), and also has substantial python overhead. This has been pullet out of GPyTorch (https://github.com/cornellius-gp/gpytorch), which implements the CG solves that you're interested in.

I'd be very interested in making this a thing in pytorch.

**Comment by iramazanli** [comment]
Conjugate Gradient descent is useful to solve system of linear equations in particular when the representation matrix of equations is positive definite. That's why it would be great fit for linear algebra module instead of torch.optim.

**Comment by NeelMishra** [comment]
Any updates?

**Comment by NeelMishra** [comment]
I have my own versions of CG and BICGSTAB that I wrote for a paper, and I use it in my own research, should I contribute it to the main PyTorch repo?

**Comment by lezcano** [comment]
If you want to follow through with this, a good way to start would be to create an external package with your code, tests, benchmarks against other implementations, and docs. This would make a strong case for it, and it would already be helpful for other practitioners that may want to use this method.

**Comment by devzhk** [comment]
I have a minimal implementation of CG and GMRES to solve linear systems with Pytorch in this [repo](https://github.com/devzhk/Pytorch-linalg). I've tested it against `scipy.sparse.linalg.cg` and `jax.scipy.sparse.linalg.cg`. If people are interested, we can work on a version that can be potentially merged into Pytorch.
